### PR TITLE
new development/deployment/release process

### LINKED ISSUE #344: new release process / CI actions / checks
We want to move to a more formal release process characterized by the following points:
* all PRs merge into `main`
* PRs don't get merged until they are finney-ready
* labels will be used to indicate that a PR is `on-devnet`, `devnet-pass`, `on-testnet`, and `testnet-pass`, respectively, and having both the `devnet-pass` and `testnet-pass` labels will be required by the CI to merge into main. 
* any issues uncovered during testing can then be discussed and tracked on the PR, reducing issue/PR bloat and making it easy to see the full life-cycle of a feature from development to deployment on finney
* we will cut releases off of main using the github releases feature, so every once in a while we cut a new release and deploy everything new in main to finney, and then that gets a semver version number + tag
* also new github actions to run our (now manual) deploy process, these will include a built in check that the spec version has been bumped to a value higher than the net we are currently deploying to. If this criterion is not met, the deploy will abort, which will prevent issues like the recent testnet forking/bricking

all work in #346 

AC:
- [x] CI check system for `devnet` / `testnet` stages (possibly using labels)
- [ ] github actions for deploys
- [ ] set up releases
- [x] PRs default to `main`
- [x] update docs / instructions / PR templates
- [ ] test that non-team-members can't fake the checklist items

### PR DESCRIPTION
**Pull request by sam0x17**
Fixes #344 
 
note that this merges into `main` because that is what determines the root CI settings for the repo once we change default branch back to `main`. May need to backport as well if there are still lingering branches on the day this is merged, but the idea of this PR is that once it is merged, all new PRs target `main`, so we shouldn't need to

will want to merge on the same day @garrett-opentensor makes the proper changes to the repo/branch protection settings, which are outlined in detail in the "branches" section of `CONTRIBUTING.md`

### CODE REVIEW

**Comment by sam0x17** [comment]
note: should probably add a PR template for the companion PRs 🤔

File: .github/workflows/label-triggers.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      - name: Check if 'breaking change' label is added
+        if: github.event.label.name == 'breaking-change'
+        uses: actions/github-script@v6
+        with:
+          script: |
+            github.rest.issues.createComment({
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: '@opentensor/cerebrum / @opentensor/gyrus / @opentensor/cortex breaking change detected! Please prepare accordingly!'
```
Comment (distributedstatemachine) [inline .github/workflows/label-triggers.yml:24]: Have we resolved permissioning issues with these alerts?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    responsibility to once again verify that the features/changes/fixes introduced by your PR
+    are functioning properly on `testnet` by interacting with the live network, if applicable.
+    If you are an external contributor, a core team member may be assigned to do this testing
+    for you but otherwise it will be your responsibility to show evidence on the PR that the
+    testing is successful. Once this has been verified, the `testnet-pass` label should be
+    added. If testing fails, the `testnet-fail` label should be added and PRs should be opened
+    reverting the change from `devnet-ready`, and then a PR should be opened merging the
+    modified `devnet` into `testnet`. These revert PRs, if they occur, _must_ be merged before
+    a new deploy can be run (TODO: enforce this via CI).
+12. After the SOP period (72 hours on testnet) has passed and the `testnet-pass` label has been
```
Comment (distributedstatemachine) [inline CONTRIBUTING.md]: is this still valid? given the new release cycle (wednesday - wednesday), I think period lasts at least a week

File: .github/workflows/testnet-labels.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+name: Tested on Testnet
+on:
+  pull_request:
+    types: [opened, labeled, unlabeled, synchronize]
```
Comment (distributedstatemachine) [inline .github/workflows/testnet-labels.yml:4]: what does `synchronise` do ?

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]

**Review by gztensor — APPROVED** [review]

File: .github/workflows/testnet-labels.yml (hunk 3 above)
Comment (sam0x17) [reply .github/workflows/testnet-labels.yml:4]: I don't understand it much but what I read is that you need it for label stuff to work 100% of the time haha

Comment (orriin) [reply .github/workflows/testnet-labels.yml:4]: "In the context of GitHub Actions, the synchronize event type for pull_request triggers a workflow whenever a pull request’s head branch is updated. This usually happens when new commits are pushed to the branch from which the pull request was created." :)

**Review by JohnReedV — APPROVED** [review]

File: .github/workflows/label-triggers.yml (hunk 1 above)
Comment (sam0x17) [reply .github/workflows/label-triggers.yml:24]: I doubt it, @garrett-opentensor will need to make sure each of those groups has permissions to receive notifications in github

File: CONTRIBUTING.md (hunk 2 above)
Comment (sam0x17) [reply CONTRIBUTING.md]: ah OK wasn't sure if that was just a new schedule vs the new actual SOP, will update

**Review by distributedstatemachine — APPROVED** [review]

**Pull request merged**